numpy/core/src/umath/matmul.c.src

```diff
@@ -267,19 +267,84 @@ NPY_NO_EXPORT void
 
 /**end repeat**/
 
+
+NPY_NO_EXPORT void
+OBJECT_matmul_inner_noblas(void *_ip1, npy_intp is1_m, npy_intp is1_n,
+                           void *_ip2, npy_intp is2_n, npy_intp is2_p,
+                           void *_op, npy_intp os_m, npy_intp os_p,
+                           npy_intp dm, npy_intp dn, npy_intp dp)                         
+{
+    char *ip1 = (char *)_ip1, *ip2 = (char *)_ip2, *op = (char *)_op;
+
+    npy_intp ib1_n = is1_n * dn;
+    npy_intp ib2_n = is2_n * dn;
+    npy_intp ib2_p = is2_p * dp;
+    npy_intp ob_p  = os_p * dp;
+
+    PyObject *tmp1, *tmp2, *tmp = NULL;
+    PyObject **tmp3;
+
+    for (npy_intp m = 0; m < dm; m++) {
+        for (npy_intp p = 0; p < dp; p++) {
+            if ( 0 == dn ) return;
+            for (npy_intp n = 0; n < dn; n++) {
+                if ((*((PyObject **)ip1) == NULL) || (*((PyObject **)ip2) == NULL)) {
+                    tmp1 = Py_False;
+                    Py_INCREF(Py_False);
+                }
+                else {
+                    tmp1 = PyNumber_Multiply(*((PyObject **)ip1), *((PyObject **)ip2));
+                    if (!tmp1) {
+                        Py_XDECREF(tmp);
+                        return;
+                    }
+                }
+                if (n == 0) {
+                    tmp = tmp1;
+                }
+                else {
+                    tmp2 = PyNumber_Add(tmp, tmp1);
+                    Py_XDECREF(tmp);
+                    Py_XDECREF(tmp1);
+                    if (!tmp2) {
+                        return;
+                    }
+                    tmp = tmp2;
```

Comment:
Can use `SETREF` here:

```C
Py_SETREF(sum_product, PyNumber_Add(sum_product, product))
Py_DECREF(product))
if (sum_product == NULL) {
    return;
}
```

Discussed but rejected at the sprint: `PyNumber_InPlaceAdd`